### PR TITLE
Upgrade React Native to 0.64

### PR DESCRIPTION
**Pull request by ceyhun**
Upgrading react-native to version 0.64.0

To test: Follow instructions on gutenberg-mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3430

## Regression Notes
1. Potential unintended areas of impact

- The whole Gutenberg editor could be affected

2. What I did to test those areas of impact (or what existing automated tests I relied on)

- Went through the [writing flow tests]( https://github.com/wordpress-mobile/test-cases/tree/master/test-cases/gutenberg/writing-flow)

3. What automated tests I added (or what prevented me from doing so)

- Gutenberg Mobile has it's own automated tests and doesn't rely on WPiOS's automated tests for now (except the limited [EditorGutenbergTests](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/WordPressUITests/Tests/EditorGutenbergTests.swift))

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by mokagio** [comment]
Bumping this to 17.7 as part of the open PRs walkthrough before starting the 17.6 code freeze because there are conflicting files and it hasn't been reviewed yet.

Ping me here if this can't wait two weeks and needs to go into 17.6 once approved. 👍

**Comment by guarani** [comment]
Hi @ceyhun 👋 

>  @buildkite buildkite/wordpress-ios — Build #629 failed (32 minutes, 53 seconds) 

Looking at https://buildkite.com/wordpress-mobile/wordpress-ios/builds/629, the UI tests for iPhone and iPad failed with errors like:

```
[14:11:57]: ▸ EditorGutenbergTests
[14:12:33]: ▸ <span class="term-fg35">    ✗ testBasicPostPublish, XCTAssertTrue failed - Page WordPressUITests.PasswordScreen is not loaded.</span>
```

Is this a known issue?

---

A couple of other things I wasn't sure about:
1. I went to the Gutenberg Mobile PR to check what to test and it says to try to run all the manual tests. I'm unsure if this means I should do the writing flow tests (as you did for Android [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3430#issuecomment-855828593)), or different manual tests (block tests for example)?
2. Was the Regression Notes section left empty intentionally?

File: Podfile [hunk 1]
```diff
@@ -416,19 +421,6 @@ end
...
-    %x(patch "#{project_root}/Pods/ReactCommon/turbomodule/core/platform/ios/RCTTurboModuleManager.mm" < "#{project_root}/patches/RCTTurboModuleManager.patch")
-
-    puts 'Patching RCTShadowView to fix nested group block - it could be removed after upgrade to 0.62'
-    %x(patch "#{project_root}/Pods/React-Core/React/Views/RCTShadowView.m" < "#{project_root}/patches/RN-RCTShadowView.patch")
-    puts 'Patching RCTActionSheet to add possibility to disable action sheet buttons -
-    it could be removed once PR with that functionality will be merged into RN'
-    %x(patch "#{project_root}/Pods/React-RCTActionSheet/RCTActionSheetManager.m" < "#{project_root}/patches/RN-RCTActionSheetManager.patch")
-    puts 'Patching RCTUIImageViewAnimated to fix a problem where images will not load when built using the iOS 14 SDK (Xcode 12) -
-    it can be removed once we upgrade Gutenberg to use RN 0.63 or later'
-    %x(patch "#{project_root}/Pods/React-RCTImage/RCTUIImageViewAnimated.m" < "#{project_root}/patches/RN-RCTUIImageViewAnimated.patch")
```
Comment (guarani) [inline Podfile:430]: Thanks for cleaning this up ❤️

Comment (guarani) [inline Podfile:430]: Thanks for cleaning this up ❤️

File: WordPress/WordPress.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -363,7 +363,7 @@
 		24B1AE3124FEC79900B9F334 /* RemoteFeatureFlagTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 24B1AE3024FEC79900B9F334 /* RemoteFeatureFlagTests.swift */; };
 		24C69A8B2612421900312D9A /* UserSettingsTests.swift in Sources */ = {isa = PBXBuildFile; fileRef = 24C69A8A2612421900312D9A /* UserSettingsTests.swift */; };
 		24C69AC22612467C00312D9A /* UserSettingsTestsObjc.m in Sources */ = {isa = PBXBuildFile; fileRef = 24C69AC12612467C00312D9A /* UserSettingsTestsObjc.m */; };
-		24CE2EB1258D687A0000C297 /* WordPressFlux in Frameworks */ = {isa = PBXBuildFile; productRef = 24CE2EB0258D687A0000C297 /* WordPressFlux */; };
```
Comment (guarani) [inline WordPress/WordPress.xcodeproj/project.pbxproj]: What are these changes for? 
Asking because they look to revert the changes from https://github.com/wordpress-mobile/WordPress-iOS/pull/16351/files, which was merged a couple of months ago. There's some context around that PR (internal reference p1619035833107300-slack-C02Q9GACF), and IIRC it has to do with changes different versions of Xcode apply to the project file.

File: Podfile [hunk 3]
```diff
@@ -117,12 +116,15 @@ def gutenberg_dependencies(options)
         'React-RCTSettings',
         'React-RCTText',
         'React-RCTVibration',
+        'React-callinvoker',
         'React-cxxreact',
         'React-jsinspector',
         'React-jsi',
         'React-jsiexecutor',
+        'React-perflogger',
+        'React-runtimeexecutor',
```
Comment (guarani) [inline Podfile:125]: Curious, are these new React modules that are part of the new version?

**Review by guarani — COMMENTED** [review]
I smoke tested the app using the App Center build and didn't see any issues. I held off manual testing because I wasn't sure which areas need testing the most.

File: Podfile (hunk 3 above)
Comment (ceyhun) [reply Podfile:125]: That's right. Normally they are included via the new [react_native_pods](https://github.com/facebook/react-native/blob/v0.64.0/scripts/react_native_pods.rb#L47-L48) script in Gutenberg demo app's Podfile [here](https://github.com/WordPress/gutenberg/pull/29118/files#diff-9a2250a2cc0b1edbfbdf9cd4edc78323b6bccd9cf02689be56a6b7b90d065d26R5). But WPiOS still needs us to define each one manually.

File: WordPress/WordPress.xcodeproj/project.pbxproj (hunk 2 above)
Comment (ceyhun) [reply WordPress/WordPress.xcodeproj/project.pbxproj]: These changes were the result of running `rake dependencies` after updating the gutenberg reference. I tried resetting `project.pbxproj` back to the `develop` version and running `rake dependencies` and the same changes appeared again. I'm using Xcode 12.4.

Next I tried moving a file up and back down in the Project navigator in Xcode and a similar diff happened as in https://github.com/wordpress-mobile/WordPress-iOS/pull/16351. I'm not sure why this is happening to us, but running `rake dependencies` now doesn't create the same diff anymore. I guess I'll go ahead and commit these changes.

**Comment by ceyhun** [comment]
Thanks for the review @guarani!

> Looking at buildkite.com/wordpress-mobile/wordpress-ios/builds/629, the UI tests for iPhone and iPad failed with errors like:
Is this a known issue?

I think it's safe to ignore Buildkite failures for now (internal references p1621980493008500-slack-CEJT26EG1 , p1623844140451700-slack-CC7L49W13).

> 1. I went to the Gutenberg Mobile PR to check what to test and it says to try to run all the manual tests. I'm unsure if this means I should do the writing flow tests (as you did for Android here), or different manual tests (block tests for example)?

After that one, there was another pass of writing flow test both on WPiOS and WPAndroid mentioned [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3430#pullrequestreview-683910055). That was done on a branch fixing WPAndroid errors targeting RN 0.64 upgrade branch. The changes for the fix were cherry-picked to the RN 0.64 upgrade branch.

> I smoke tested the app using the App Center build and didn't see any issues. I held off manual testing because I wasn't sure which areas need testing the most.

Thanks again! I think that should be enough for now and other than that I guess it would take a lot of time to go through every manual test and we may have to rely on automated ones and the future release tests we'll have. 

> 2. Was the Regression Notes section left empty intentionally?

I thought the regression notes didn't really fit well with a big change like a React Native upgrade it's really hard to tell. I guess the whole Gutenberg editor can be affected by these changes and we didn't mostly rely on WPiOS's automated tests for catching any regressions. I'll add this in there.

File: Podfile (hunk 3 above)
Comment (guarani) [reply Podfile:125]: Thanks for explaining 👍

File: WordPress/WordPress.xcodeproj/project.pbxproj (hunk 2 above)
Comment (guarani) [reply WordPress/WordPress.xcodeproj/project.pbxproj]: Ok, they look like harmless changes, I was only concerned that they would undo https://github.com/wordpress-mobile/WordPress-iOS/pull/16351/files, which was also made with Xcode 12.4.
Since the changes have now been removed from this PR, I'll mark this done.

**Review by guarani — APPROVED** [review]
> I think it's safe to ignore Buildkite failures for now (internal references p1621980493008500-slack-CEJT26EG1 , p1623844140451700-slack-CC7L49W13).

Ah yes, I forgot that buildkite errors aren't blockers.

> After that one, there was another pass of writing flow test both on WPiOS and WPAndroid mentioned [here](https://github.com/wordpress-mobile/gutenberg-mobile/pull/3430#pullrequestreview-683910055). That was done on a branch fixing WPAndroid errors targeting RN 0.64 upgrade branch. The changes for the fix were cherry-picked to the RN 0.64 upgrade branch.

Sounds good!

> Thanks again! I think that should be enough for now and other than that I guess it would take a lot of time to go through every manual test and we may have to rely on automated ones and the future release tests we'll have.

I think that makes sense.

> I thought the regression notes didn't really fit well with a big change like a React Native upgrade it's really hard to tell. I guess the whole Gutenberg editor can be affected by these changes and we didn't mostly rely on WPiOS's automated tests for catching any regressions. I'll add this in there.

Cool, makes sense. I just noticed you added "Gutenberg Mobile has it's own automated tests and doesn't rely on WPiOS's automated tests for now" above. My understanding was that [`EditorGutenbergTests`](https://github.com/wordpress-mobile/WordPress-iOS/blob/develop/WordPress/WordPressUITests/Tests/EditorGutenbergTests.swift) is relied upon to some extent to catch editor regressions – even though it only really does a little bit of smoke testing of the editor. Not sure if that's what you meant, just wanted to mention it.

Looking forward to using 0.64! 🕺 Thanks for all the work here! 🙇

**Comment by ceyhun** [comment]
> Cool, makes sense. I just noticed you added "Gutenberg Mobile has it's own automated tests and doesn't rely on WPiOS's automated tests for now" above. My understanding was that EditorGutenbergTests is relied upon to some extent to catch editor regressions – even though it only really does a little bit of smoke testing of the editor. Not sure if that's what you meant, just wanted to mention it.

I keep forgetting about that one 😄 But yeah, I guess we'll not be adding any more automated tests other than that for Gutenberg for the time being. Thanks a lot for the review 🙇

**Comment by mokagio** [comment]
🎉
